modules/apps/callbacks/ibc_middleware.go

```diff
@@ -0,0 +1,243 @@
+	"github.com/cosmos/ibc-go/v7/modules/apps/callbacks/types"
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v7/modules/core/05-port/types"
+	ibcerrors "github.com/cosmos/ibc-go/v7/modules/core/errors"
+	ibcexported "github.com/cosmos/ibc-go/v7/modules/core/exported"
+)
+
+var (
+	_ porttypes.Middleware            = (*IBCMiddleware)(nil)
+	_ porttypes.PacketDataUnmarshaler = (*IBCMiddleware)(nil)
+)
+
+// IBCMiddleware implements the ICS26 callbacks for the ibc-callbacks middleware given
+// the underlying application.
+type IBCMiddleware struct {
+	app     types.PacketUnmarshalerIBCModule
+	channel porttypes.ICS4Wrapper
+
+	contractKeeper types.ContractKeeper
+}
+
+// NewIBCMiddleware creates a new IBCMiddlware given the keeper and underlying application
+func NewIBCMiddleware(
+	app porttypes.IBCModule,
+	channel porttypes.ICS4Wrapper,
+	contractKeeper types.ContractKeeper,
+) IBCMiddleware {
+	packetUnmarshalerApp, ok := app.(types.PacketUnmarshalerIBCModule)
+	if !ok {
+		panic("underlying application does not implement PacketDataUnmarshaler")
+	}
+	return IBCMiddleware{
+		app:            packetUnmarshalerApp,
+		channel:        channel,
+		contractKeeper: contractKeeper,
+	}
+}
+
+// UnmarshalPacketData defers to the underlying app to unmarshal the packet data.
+func (im IBCMiddleware) UnmarshalPacketData(bz []byte) (interface{}, error) {
+	return im.app.UnmarshalPacketData(bz)
+}
+
+// OnAcknowledgementPacket implements source callbacks for acknowledgement packets.
+// It defers to the underlying application and then calls the contract callback.
+// If the contract callback fails (within the gas limit), state changes are reverted.
+func (im IBCMiddleware) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	// we first call the underlying app to handle the acknowledgement
+	appResult := im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	if appResult != nil {
+		return appResult
+	}
+
+	callbackData, err := types.GetCallbackData(im.app, packet, ctx.GasMeter().GasRemaining())
+	if err != nil {
+		types.EmitSourceCallbackEvent(ctx, packet, types.CallbackTypeAcknowledgement, callbackData, err)
+		return appResult
+	}
+	if callbackData.SrcContractAddr == "" {
+		return appResult
+	}
+
+	var ack channeltypes.Acknowledgement
+	if err := channeltypes.SubModuleCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
+		err = errorsmod.Wrapf(ibcerrors.ErrUnknownRequest, "cannot unmarshal callback packet acknowledgement: %v", err)
+		types.EmitSourceCallbackEvent(ctx, packet, types.CallbackTypeAcknowledgement, callbackData, err)
+		return appResult
+	}
+
+	cachedCtx, writeFn := ctx.CacheContext()
+	cachedCtx = cachedCtx.WithGasMeter(sdk.NewGasMeter(callbackData.GasLimit))
```

Comment:
we also need to catch the outofgas panic if the cached ctx runs out of gas

Reply:
done